lib/xenops_server.ml

```diff
@@ -1328,15 +1336,46 @@ let rec immediate_operation dbg id op =
 			Xenops_task.run task
 		) ();
 	match Xenops_task.get_state task with
-		| Task.Pending _ -> assert false
-		| Task.Completed _ -> ()
-		| Task.Failed e ->
-			let e = e |> Exception.exnty_of_rpc |> exn_of_exnty in
-			raise e
+	| Task.Pending _ -> assert false
+	| Task.Completed _ -> ()
+	| Task.Failed e ->
+		let e = e |> Exception.exnty_of_rpc |> exn_of_exnty in
+		raise e
+
+and queue_operation_int dbg id op =
+	let task = Xenops_task.add tasks dbg (let r = ref None in fun t -> perform ~result:r op t; !r) in
+	let tag = if uses_mxgpu id then "mxgpu" else id in
+	Redirector.push Redirector.default tag (op, task);
+	task
+
+and queue_operation dbg id op =
+	debug "queue_operation: %s" dbg;
+	let task = queue_operation_int dbg id op in
+	Xenops_task.id_of_handle task
+
+and queue_operation_and_wait dbg id op =
+	debug "queue_operation_and_wait: %s" dbg;
+	let from = Updates.last_id dbg updates in
+	let task = queue_operation_int dbg id op in
+	let task_id = Xenops_task.id_of_handle task in
+	event_wait updates task ~from 1200.0 (task_finished_p task_id) |> ignore;
+	task
+
+(* CA-254911: delay the cleanups checks and move them to a different
+ * thread to try and mitigate the risk of stack overflow due to 
+ * exceptional issues *)
+and deferred_operation dbg id op =
+	let task = queue_operation_and_wait dbg id op in
+	match Xenops_task.get_state task with
```

Comment:
Only the 
```ocaml
match Xenops_task.get_state task with
	| Task.Pending _ -> assert false
	| Task.Completed _ -> ()
	| Task.Failed e ->
		let e = e |> Exception.exnty_of_rpc |> exn_of_exnty in
		raise e
```
is common. I am not sure if it is worth extracting it to a function, more often than not the task state is dealt with in a specific way, and the common helper would be just for these two functions.